### PR TITLE
Fix Windows cotire build

### PR DESCRIPTION
**Pull request by HubertBalcerzak**
This allows building Kratos with cotire on windows

### CODE REVIEW

File: scripts/build/appveyor/configure_appveyor_VS2017.bat [hunk 1]
```diff
@@ -11,7 +11,7 @@ set CC=cl.exe
 set CXX=cl.exe
 
 rem Set variables
-set KRATOS_BUILD_TYPE=Release
+set KRATOS_BUILD_TYPE=Debug
```
Comment (philbucher) [inline scripts/build/appveyor/configure_appveyor_VS2017.bat:14]: is it worth updating appveyor?
I would not touch it any more

**Comment by philbucher** [comment]
at the end of the build it says: 
~~~
  -- Install configuration: "Debug"
  -- Deleting: D:/a/Kratos/Kratos/bin/Debug/KratosMultiphysics
  -- Deleting: D:/a/Kratos/Kratos/bin/Debug/libs
~~~

maybe thats why it cant find Kratos when trying to run the tests

**Comment by roigcarlo** [comment]
Nono, that's normal, it says that because prior to install we decided to delete the folder. Files get installed after, but at some point we also disable the messages.

The problem is that the env variables defined in the previous stage are not available here.

File: scripts/build/appveyor/configure_appveyor_VS2017.bat (hunk 1 above)
Comment (roigcarlo) [reply scripts/build/appveyor/configure_appveyor_VS2017.bat:14]: In any case, lets keep things consistent, its just adding a couple of lines...

**Comment by philbucher** [comment]
yay it works
well, tests fail :)
maybe too slow in debug?

**Comment by philbucher** [comment]
@GuillermoCasas there is a test in the SwimmingDEM app that uses the non-standard python module `mpmath`. Can you please add a skiptest if this module is not available?
Thx

**Comment by roigcarlo** [comment]
Yep, to slow, so... we are a little f*****. Either we increase the runtime for the test, or we compile in Release, and the latter is not an option because then we run out of memory in the machine (cof cof @loumalouomega templates cof cof).

**Comment by philbucher** [comment]
> Yep, to slow, so... we are a little f*****. Either we increase the runtime for the test, or we compile in Release, and the latter is not an option because then we run out of memory in the machine (cof cof @loumalouomega templates cof cof).

I try now without contact and in release. I mean, the problem with the templates is not precisely new...

**Comment by roigcarlo** [comment]
It runs out of memory at core level...

**Comment by philbucher** [comment]
oh ok
is this the same as #6284 ?

**Comment by roigcarlo** [comment]
Yep...

**Comment by philbucher** [comment]
@roigcarlo do you know which files are the problem?
Can we test this somehow?

**Comment by philbucher** [comment]
Ok it fails even before, in the core-Core :/
Why does it need so much memory? I mena, in linux it works...

release is compiled with `O2` right?
Debug is `O0` and `-g` right?
Maybe we can do a custom build with `O0` or `O1` and without `-g`

what do you think @roigcarlo ?

**Comment by roigcarlo** [comment]
> Why does it need so much memory? I mena, in linux it works...

MSVC is trash.

> Maybe we can do a custom build with O0 or O1 and without -g

For linux, what you say is exactly what the custom build mode does. Maybe we could increase the O0 to O1 to see if the github machines can handle it.

For win... we can put O1, but I don't know if that is a level of optimization valid for MSVC... as was saying the compiler is quite special and unique.

**Comment by loumalouomega** [comment]
> 
> 
> > Why does it need so much memory? I mena, in linux it works...
> 
> MSVC is trash.
> 
> > Maybe we can do a custom build with O0 or O1 and without -g
> 
> For linux, what you say is exactly what the custom build mode does. Maybe we could increase the O0 to O1 to see if the github machines can handle it.
> 
> For win... we can put O1, but I don't know if that is a level of optimization valid for MSVC... as was saying the compiler is quite special and unique.

Join MinGW dark side...

**Comment by philbucher** [comment]
https://docs.microsoft.com/en-us/cpp/build/reference/o-options-optimize-code?view=vs-2019

 I am not the expert but I think we could try some things...?

**Comment by loumalouomega** [comment]
> 
> 
> https://docs.microsoft.com/en-us/cpp/build/reference/o-options-optimize-code?view=vs-2019
> 
> I am not the expert but I think we could try some things...?

imagen

**Comment by roigcarlo** [comment]
We can, but be aware that -O1 in gcc != /O1 in msvc.

If you look at: https://www.rapidtables.com/code/linux/gcc/gcc-o.html It states that O1 decreases both execution and code usage, while increasing memory and compile time.

MSVC /O1 only reduces the binary size, but does nothing with the execution time, while -O2 (whch is the only other code performance optimization level btw..., with maybe the exception of /Ox, but I don't really know what they do there) behaves like -O3 of gcc (at least it tries to...)

**Comment by philbucher** [comment]
can we just try those flags one by one to see if sth works

I don't know though how to set them :)

**Comment by roigcarlo** [comment]
> I don't know though how to set them :)

Add `-DCMAKE_CXX_FLAGS="/Ox"  ^` to the compilation options and the build type to `Custom`.

Feel free to try :)

**Comment by philbucher** [comment]
@roigcarlo is LTO on or off?

also this might be interesting, but then again I don't know how to set those things in CMake :) https://github.com/tensorflow/tensorflow/issues/11096

**Comment by roigcarlo** [comment]
Off

**Comment by philbucher** [comment]
> > https://docs.microsoft.com/en-us/cpp/build/reference/o-options-optimize-code?view=vs-2019
> > I am not the expert but I think we could try some things...?
> 
> imagen

image

**Comment by philbucher** [comment]
now it is able to build and also run the tests in a decent amount of time :D

if fails overall because some tests are failing

**Comment by philbucher** [comment]
Tests are failing in
DEM @KratosMultiphysics/dem 
SwimmingDEM @KratosMultiphysics/dem @GuillermoCasas 
StructuralMechanics @philbucher (my test)

Please take a look and fix those tests such that we can merge this :)

**Comment by philbucher** [comment]
@loumalouomega also the contact tests seem to crash in Win

**Comment by philbucher** [comment]
@KratosMultiphysics/technical-committee just as an idea, we could for now disable the failing apps, merge this PR and then add the apps back in separate PRs
They are still ensured to compile in Win because they are in appveyor
this way this PR won't hang ...

**Comment by roigcarlo** [comment]
Agree

**Comment by philbucher** [comment]
Now it passes :D

**Comment by HubertBalcerzak** [comment]
@philbucher Is there anything still blocking this PR?

**Review by philbucher — APPROVED** [review]
To me this can/should go in now, none had any comments and the longer we wait the higher the chances are that some test will break

For the apps I disabled, the are still being compiled in Appveyor so nothing is lost on that end. We should fix the tests and add the apps back here asap though

The night CI that is failing is because there are some tests failing in the nightly suites, this is a known issue, and is being addressed constantly 

Thanks for the efforts @HubertBalcerzak @roigcarlo

**Comment by roigcarlo** [comment]
@philbucher Seems that windows nightly is mandatory so we cannot merge :S

**Comment by philbucher** [comment]
Hm interesting/strange
I will check next week, probably we need to rename sth

**Comment by philbucher** [comment]
now it should work once the CI has passed

**Comment by philbucher** [comment]
yep now you can merge

**Comment by roigcarlo** [comment]
Ok let's do it

**Comment by philbucher** [comment]
:tada: :tada: :tada:

**Pull request merged**